### PR TITLE
Authenticate with ID token

### PR DESCRIPTION
**Pull request by byewokko**
## ID token
- ID token now properly signed with ES256
- In provisioning mode, OIDC service creates EC private key if not exists

---

## Authentication
- **`BREAKING`** Seacat API (=private container) now authenticates only via ID token (instead of access token)
  - Set `[seacat:api] _allow_access_token_auth=yes` to enable fallback authentication with access token. This is for dev purposes only!
- **`BREAKING`** All introspection endpoints now output ID token (instead of access token)
- To allow authentication via access tokens for Seacat API, the recommended way is to set up `auth_request` on the API location in nginx:
```nginx
location /seacat/api/seacat_auth {
	auth_request       /_oauth2_introspect;
	auth_request_set   $authorization $upstream_http_authorization;
	proxy_set_header   Authorization $authorization;
	rewrite    ^/seacat/api/seacat_auth/(.*) /$1 break;
	proxy_pass http://localhost:8082;
}
```
This will trade the incoming access token header for ID token.

---

## `GET /openidconnect/public_keys`
- returns the app's public keys in JWK JSON format, for example
```json
{
	"keys": [
		{
			"kty": "EC",
			"kid": "moDZIlfjvAAAUcQs9D72XOTToUiyxwKw43MhIRLBD4w",
			"crv": "P-256",
			"x": "IBsxa21kSTVAr-RZwGXHCRN2moSfqR6xwlhRwy01ZPE",
			"y": "tqaS-ByGTLygVln-1xShWYHrg068aq2vFb-IngW7VQQ"
		}
	]
}
```
- only one key is supported for now
- add `?format=pem` to query to return the keys in PEM format, for example
```json
{
	"moDZIlfjvAAAUcQs9D72XOTToUiyxwKw43MhIRLBD4w": "-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEIBsxa21kSTVAr+RZwGXHCRN2moSf\nqR6xwlhRwy01ZPG2ppL4HIZMvKBWWf7XFKFZgeuDTrxqra8Vv4ieBbtVBA==\n-----END PUBLIC KEY-----\n"
}
```
- return format inspired by Google's https://www.googleapis.com/oauth2/v1/certs and https://www.googleapis.com/oauth2/v3/certs

### CODE REVIEW

File: seacatauth/openidconnect/__init__.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from .handler.userinfo import UserInfoHandler
 from .handler.introspect import TokenIntrospectionHandler
 from .handler.session import SessionHandler
+from .handler.keys import KeysHandler
```
Comment (ateska) [inline seacatauth/openidconnect/__init__.py:10]: `PublicKeysHandler`

File: seacatauth/openidconnect/handler/keys.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		# Public endpoints
+		web_app_public = app.PublicWebContainer.WebApp
+		web_app_public.router.add_get("/openidconnect/keys", self.keys)
+
+
+	async def keys(self, request):
+		"""
+		Return public keys that the client can use for token verification.
+
+		Specify ?pem in query to return the keys in PEM format.
```
Comment (ateska) [inline seacatauth/openidconnect/handler/keys.py:32]: `?format=pem`

File: seacatauth/openidconnect/handler/userinfo.py [hunk 3]
```diff
@@ -55,8 +55,6 @@ async def userinfo(self, request):
 			session,
 			tenant=request.query.get("tenant", "*")
 		)
-		if userinfo["result"] == "CREDENTIALS-NOT-FOUND":
-			return self.error_response("invalid_credentials", "Invalid credentials.")
```
Comment (ateska) [inline seacatauth/openidconnect/handler/userinfo.py:59]: Is there a check for incorrect userinfo then?

File: seacatauth/openidconnect/service.py [hunk 4]
```diff
@@ -61,13 +61,71 @@ def __init__(self, app, service_name="seacatauth.OpenIdConnectService"):
...
+		self.APIAllowAccessToken = asab.Config.getboolean("seacat:api", "_allow_access_token_auth")
+
 		self.App.PubSub.subscribe("Application.tick/60!", self._on_tick)
 
 
 	async def _on_tick(self, event_name):
 		await self.delete_expired_authorization_codes()
 
 
+	def _prepare_private_key(self):
```
Comment (ateska) [inline seacatauth/openidconnect/service.py:75]: `_load_private_key()`

File: seacatauth/openidconnect/service.py [hunk 5]
```diff
@@ -61,13 +61,71 @@ def __init__(self, app, service_name="seacatauth.OpenIdConnectService"):
...
+			L.log(
+				asab.LOG_NOTICE,
+				"OpenIDConnect private key file not specified. Defaulting to '{}'.".format(private_key_path)
+			)
+
+		if os.path.isfile(private_key_path):
+			with open(private_key_path, "rb") as f:
+				private_key = jwcrypto.jwk.JWK.from_pem(f.read())
+		else:
+			# Generate a new private key
```
Comment (ateska) [inline seacatauth/openidconnect/service.py:98]: Only in provisioning mode.

File: seacatauth/openidconnect/service.py [hunk 6]
```diff
@@ -128,17 +187,53 @@ async def get_session_from_bearer_token(self, bearer_token: str):
...
+
+		token_string = match.group(1)
+
+		try:
+			token = jwcrypto.jwt.JWT(jwt=token_string, key=self.PrivateKey)
+		except jwcrypto.jwt.JWTExpired:
+			L.warning("ID Token expired")
+			return None
+		except jwcrypto.jws.InvalidJWSSignature:
+			L.warning("Invalid ID Token signature")
```
Comment (ateska) [inline seacatauth/openidconnect/service.py:204]: "ID Token" vs "ID token" ... please unify and not only here.

**Review by ateska — CHANGES_REQUESTED** [review]

File: seacatauth/openidconnect/handler/userinfo.py (hunk 3 above)
Comment (byewokko) [reply seacatauth/openidconnect/handler/userinfo.py:59]: at this moment userinfo is always created from session data. and session cannot be created with invalid credentials.

so this check is redundant, at least for now.

File: seacatauth/openidconnect/handler/session.py [hunk 7]
```diff
@@ -4,6 +4,7 @@
 import aiohttp.web
 
 #
+from seacatauth.generic import get_bearer_token_value
```
Comment (ateska) [inline seacatauth/openidconnect/handler/session.py:7]: Relative import.

File: seacatauth/openidconnect/handler/userinfo.py [hunk 8]
```diff
@@ -7,6 +7,7 @@
 import asab.web.rest
 
 #
+from seacatauth.generic import get_bearer_token_value
```
Comment (ateska) [inline seacatauth/openidconnect/handler/userinfo.py:10]: Relative imports

**Review by ateska — CHANGES_REQUESTED** [review]
Package imports should be relative. Then mergeable.

**Pull request merged**